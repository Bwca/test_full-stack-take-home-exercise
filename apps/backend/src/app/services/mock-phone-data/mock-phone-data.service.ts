import { Injectable } from '@nestjs/common';

import { PhoneEntry } from '@full-stack-take-home-exercise/models';

import { PhoneDataService } from '../phone-data.service';
import { DB } from './db';

@Injectable()
export class MockPhoneDataService implements PhoneDataService {
  public getEntryByPhoneNumber(number: string): PhoneEntry {
    const result = [...this.loadDb]
      .sort((a, b) => b.prefix - a.prefix)
      .find(({ prefix }) => number.startsWith(String(prefix)));

    if (!result) {
      throw new Error('Could not locate a phone entry for the entered number!');
    }
    return result;
  }

  private get loadDb(): Readonly<Array<Readonly<PhoneEntry>>> {
    return Object.freeze([...DB]);
  }
}
