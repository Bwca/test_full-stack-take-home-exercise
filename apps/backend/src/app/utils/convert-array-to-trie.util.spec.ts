import { DB } from '../services/mock-phone-data/db';
import { convertArrayToTrie } from './convert-array-to-trie.util';

describe('Tests for trie converter', () => {
  it('should convert DB to trie', () => {
    // Arrange
    const db = [...DB];

    // Act
    const result = convertArrayToTrie(db);

    // Assert
    expect(result).toMatchSnapshot();
  });
});
