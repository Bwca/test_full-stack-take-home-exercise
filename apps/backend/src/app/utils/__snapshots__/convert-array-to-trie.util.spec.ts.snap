// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for trie converter should convert DB to trie 1`] = `
Object {
  "children": Object {
    "1": Object {
      "children": Object {
        "2": Object {
          "children": Object {
            "0": Object {
              "children": Object {
                "4": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1204,
                    "region": "Manitoba",
                  },
                },
              },
            },
            "2": Object {
              "children": Object {
                "6": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1226,
                    "region": "Ontario",
                  },
                },
              },
            },
            "3": Object {
              "children": Object {
                "6": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1236,
                    "region": "British Columbia",
                  },
                },
              },
            },
            "4": Object {
              "children": Object {
                "9": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1249,
                    "region": "Ontario",
                  },
                },
              },
            },
            "5": Object {
              "children": Object {
                "0": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1250,
                    "region": "British Columbia",
                  },
                },
              },
            },
            "8": Object {
              "children": Object {
                "9": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1289,
                    "region": "Ontario",
                  },
                },
              },
            },
          },
        },
        "3": Object {
          "children": Object {
            "0": Object {
              "children": Object {
                "6": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1306,
                    "region": "Saskatchewan",
                  },
                },
              },
            },
            "4": Object {
              "children": Object {
                "3": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1343,
                    "region": "Ontario",
                  },
                },
              },
            },
            "6": Object {
              "children": Object {
                "5": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1365,
                    "region": "Ontario",
                  },
                },
              },
            },
          },
        },
        "4": Object {
          "children": Object {
            "0": Object {
              "children": Object {
                "3": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1403,
                    "region": "Alberta",
                  },
                },
              },
            },
            "1": Object {
              "children": Object {
                "6": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1416,
                    "region": "Ontario",
                  },
                },
                "8": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1418,
                    "region": "Québec",
                  },
                },
              },
            },
            "3": Object {
              "children": Object {
                "1": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1431,
                    "region": "Manitoba",
                  },
                },
                "7": Object {
                  "children": Object {
                    "3": Object {
                      "children": Object {
                        "2": Object {
                          "children": Object {
                            "9": Object {
                              "children": Object {},
                              "value": Object {
                                "country": "Canada",
                                "country_code": 1,
                                "operator": "Lucky Mobile",
                                "prefix": 1437329,
                                "region": "Ontario",
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1437,
                    "region": "Ontario",
                  },
                },
                "8": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1438,
                    "region": "Québec",
                  },
                },
              },
            },
            "5": Object {
              "children": Object {
                "0": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1450,
                    "region": "Québec",
                  },
                },
              },
            },
          },
        },
        "5": Object {
          "children": Object {
            "0": Object {
              "children": Object {
                "6": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1506,
                    "region": "New Brunswick",
                  },
                },
              },
            },
            "1": Object {
              "children": Object {
                "4": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1514,
                    "region": "Québec",
                  },
                },
                "9": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1519,
                    "region": "Ontario",
                  },
                },
              },
            },
            "7": Object {
              "children": Object {
                "9": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1579,
                    "region": "Québec",
                  },
                },
              },
            },
            "8": Object {
              "children": Object {
                "1": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1581,
                    "region": "Québec",
                  },
                },
                "7": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1587,
                    "region": "Alberta",
                  },
                },
              },
            },
          },
        },
        "6": Object {
          "children": Object {
            "0": Object {
              "children": Object {
                "4": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1604,
                    "region": "British Columbia",
                  },
                },
              },
            },
            "1": Object {
              "children": Object {
                "3": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1613,
                    "region": "Ontario",
                  },
                },
              },
            },
            "3": Object {
              "children": Object {
                "9": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1639,
                    "region": "Saskatchewan",
                  },
                },
              },
            },
            "4": Object {
              "children": Object {
                "7": Object {
                  "children": Object {
                    "9": Object {
                      "children": Object {
                        "9": Object {
                          "children": Object {
                            "3": Object {
                              "children": Object {},
                              "value": Object {
                                "country": "Canada",
                                "country_code": 1,
                                "operator": "Telus",
                                "prefix": 1647993,
                                "region": "Ontario",
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1647,
                    "region": "Ontario",
                  },
                },
              },
            },
          },
        },
        "7": Object {
          "children": Object {
            "0": Object {
              "children": Object {
                "5": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1705,
                    "region": "Ontario",
                  },
                },
                "9": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1709,
                    "region": "Newfoundland and Labrador",
                  },
                },
              },
            },
            "7": Object {
              "children": Object {
                "8": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1778,
                    "region": "British Columbia",
                  },
                },
              },
            },
            "8": Object {
              "children": Object {
                "0": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1780,
                    "region": "Alberta",
                  },
                },
              },
            },
          },
        },
        "8": Object {
          "children": Object {
            "0": Object {
              "children": Object {
                "7": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1807,
                    "region": "Ontario",
                  },
                },
              },
            },
            "1": Object {
              "children": Object {
                "9": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1819,
                    "region": "Québec",
                  },
                },
              },
            },
            "6": Object {
              "children": Object {
                "7": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1867,
                    "region": "Northwest Territories",
                  },
                },
              },
            },
            "7": Object {
              "children": Object {
                "3": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1873,
                    "region": "Québec",
                  },
                },
              },
            },
          },
        },
        "9": Object {
          "children": Object {
            "0": Object {
              "children": Object {
                "2": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1902,
                    "region": "Prince Edward Island",
                  },
                },
                "5": Object {
                  "children": Object {},
                  "value": Object {
                    "country": "Canada",
                    "country_code": 1,
                    "operator": null,
                    "prefix": 1905,
                    "region": "Ontario",
                  },
                },
              },
            },
          },
        },
      },
      "value": Object {
        "country": "Canada",
        "country_code": 1,
        "operator": null,
        "prefix": 1,
        "region": null,
      },
    },
  },
}
`;
